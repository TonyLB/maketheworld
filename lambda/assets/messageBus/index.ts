import {
    MessageBus,
    isFetchLibraryAPIMessage,
    isFetchAssetAPIMessage,
    isUploadResponseMessage
} from "./baseClasses"
import fetchLibraryMessage from "../fetchLibrary"
import fetchAssetMessage from "../fetch"
import { uploadResponseMessage } from "../upload/uploadResponse"

export const messageBus = new MessageBus()

messageBus.subscribe({
    tag: 'FetchLibrary',
    priority: 5,
    filter: isFetchLibraryAPIMessage,
    callback: fetchLibraryMessage
})
messageBus.subscribe({
    tag: 'FetchAsset',
    priority: 5,
    filter: isFetchAssetAPIMessage,
    callback: fetchAssetMessage
})
messageBus.subscribe({
    tag: 'UploadResponse',
    priority: 8,
    filter: isUploadResponseMessage,
    callback: uploadResponseMessage
})

export default messageBus