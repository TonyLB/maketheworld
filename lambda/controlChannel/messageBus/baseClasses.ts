export type PublishMessageBase = {
    type: 'PublishMessage';
    targets: string[];
}

//
// TODO: Figure out how to abstract copied typescript constraints on messages
// into a central library shared between lambdas and front-end
//
type TaggedText = {
    tag: 'String';
    value: string;
}

type TaggedLineBreak = {
    tag: 'LineBreak';
}

type TaggedActionLink = {
    targetTag: 'Action';
    toAssetId: string;
    toAction: string;
}

type TaggedFeatureLink = {
    targetTag: 'Feature';
    toFeatureId: string;
}

type TaggedLinkPayload = TaggedActionLink | TaggedFeatureLink

export type TaggedLink = {
    tag: 'Link',
    RoomId: string;
    text: string;
} & TaggedLinkPayload

export type TaggedMessageContent = TaggedLink | TaggedText | TaggedLineBreak;

export type PublishWorldMessage = PublishMessageBase & {
    displayProtocol: 'WorldMessage';
    message: TaggedMessageContent[];
}

export type LegalCharacterColor = 'blue' | 'pink' | 'purple' | 'green' | 'grey'

type MessageCharacterInfo = {
    characterId: string;
    name: string;
    color: LegalCharacterColor;
}

export type PublishSpeechMessage = PublishMessageBase & MessageCharacterInfo & {
    displayProtocol: 'SayMessage';
    message: TaggedMessageContent[];
}

export type PublishNarrateMessage = PublishMessageBase & MessageCharacterInfo & {
    displayProtocol: 'NarrateMessage';
    message: TaggedMessageContent[];
}

export type PublishOutOfCharacterMessage = PublishMessageBase & MessageCharacterInfo & {
    displayProtocol: 'OOCMessage';
    message: TaggedMessageContent[];
}

export type PublishMessage = PublishWorldMessage | PublishSpeechMessage | PublishNarrateMessage | PublishOutOfCharacterMessage

export type ReturnValueMessage = {
    type: 'ReturnValue';
    body: Record<string, any>;
}

export type MessageType = PublishMessage | ReturnValueMessage

export const isPublishMessage = (prop: MessageType): prop is PublishMessage => (prop.type === 'PublishMessage')
export const isWorldMessage = (prop: PublishMessage): prop is PublishWorldMessage => (prop.displayProtocol === 'WorldMessage')
export const isCharacterMessage = (prop: PublishMessage): prop is (PublishSpeechMessage | PublishNarrateMessage | PublishOutOfCharacterMessage) => (['SayMessage', 'NarrateMessage', 'OOCMessage'].includes(prop.displayProtocol))

export const isReturnValueMessage = (prop: MessageType): prop is ReturnValueMessage => (prop.type === 'ReturnValue')
