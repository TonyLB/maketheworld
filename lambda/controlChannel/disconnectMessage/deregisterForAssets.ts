import { DisconnectMessage, MessageBus } from "../messageBus/baseClasses"
import { assetDB } from "@tonylb/mtw-utilities/dist/dynamoDB"

export const discconnectForAssetsMessage = async ({ payloads, messageBus }: { payloads: DisconnectMessage[], messageBus: MessageBus }): Promise<void> => {

    await Promise.all([
        payloads.map(({ connectionId }) => {
            assetDB.deleteItem({
                AssetId: `CONNECTION#${connectionId}`,
                DataCategory: 'Meta::Connection'
            })
        }),
        assetDB.optimisticUpdate({
            key: {
                AssetId: 'Global',
                DataCategory: 'Connections'    
            },
            updateKeys: ['connections'],
            updateReducer: (draft: { connections?: Record<string, string> }) => {
                payloads.forEach((payload) => {
                    if (draft.connections === undefined) {
                        draft.connections = {}
                    }
                    delete draft.connections[payload.connectionId]
                })
            },
        })
    ])

}

export default discconnectForAssetsMessage
