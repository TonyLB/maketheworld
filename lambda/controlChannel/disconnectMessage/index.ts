import { DisconnectMessage, MessageBus } from "../messageBus/baseClasses"

import { connectionDB, ephemeraDB, exponentialBackoffWrapper, multiTableTransactWrite } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { marshall } from "@aws-sdk/util-dynamodb"
import { splitType } from "@tonylb/mtw-utilities/dist/types"
import messageBus from "../messageBus"
import internalCache from "../internalCache"

type RoomCharacterActive = {
    EphemeraId: string;
    Color?: string;
    ConnectionIds: string[];
    fileURL?: string;
    Name: string;
}

const atomicallyRemoveCharacterAdjacency = async (connectionId, characterId) => {
    return exponentialBackoffWrapper(async () => {
        const [connectionFetch, characterFetch] = await Promise.all([
            connectionDB.getItem<{ connections: string[] }>({
                ConnectionId: `CHARACTER#${characterId}`,
                DataCategory: 'Meta::Character',
                ProjectionFields: ['connections']
            }),
            internalCache.CharacterMeta.get(characterId)
        ])
        const { connections: currentConnections } = connectionFetch || {}
        if (!currentConnections) {
            return
        }
        const { RoomId, Name, fileURL, Color } = characterFetch || {}

        const currentActiveCharacters = await internalCache.RoomCharacterList.get(characterId)

        const remainingConnections = currentConnections.filter((value) => (value !== connectionId))

        const remainingCharacters = [
            ...(currentActiveCharacters || []).filter(({ EphemeraId }) => (EphemeraId !== `CHARACTER#${characterId}`)),
            ...((remainingConnections.length > 0)
                ? [{
                    ...characterFetch,
                    ConnectionIds: remainingConnections
                }]
                : []
            )
        ]
        const adjustMeta = remainingConnections.length > 0
            ? [{
                Update: {
                    TableName: 'Connections',
                    Key: marshall({
                        ConnectionId: `CHARACTER#${characterId}`,
                        DataCategory: 'Meta::Character'
                    }),
                    UpdateExpression: 'SET connections = :newConnections',
                    ExpressionAttributeValues: marshall({
                        ':newConnections': remainingConnections,
                        ':oldConnections': currentConnections
                    }),
                    ConditionExpression: 'connections = :oldConnections'
                }
            }]
            : [{
                Delete: {
                    TableName: 'Connections',
                    Key: marshall({
                        ConnectionId: `CHARACTER#${characterId}`,
                        DataCategory: 'Meta::Character'
                    }),
                }
            }]
        await multiTableTransactWrite([{
            Delete: {
                TableName: 'Connections',
                Key: marshall({
                    ConnectionId: `CONNECTION#${connectionId}`,
                    DataCategory: `CHARACTER#${characterId}`
                })
            }
        },
        ...adjustMeta,
        {
            Update: {
                TableName: 'Ephemera',
                Key: marshall({
                    EphemeraId: `ROOM#${RoomId}`,
                    DataCategory: 'Meta::Room'
                }),
                UpdateExpression: 'SET activeCharacters = :newCharacters',
                ExpressionAttributeValues: marshall({
                    ':newCharacters': remainingCharacters,
                    ':oldCharacters': currentActiveCharacters
                }),
                ConditionExpression: 'activeCharacters = :oldCharacters'
            }
        }])
        if (remainingConnections.length === 0) {
            messageBus.send({
                type: 'EphemeraUpdate',
                updates: [{
                    type: 'CharacterInPlay',
                    CharacterId: characterId,
                    Name: Name || '',
                    Connected: false,
                    RoomId: RoomId || '',
                    fileURL: fileURL || '',
                    Color: Color || 'grey'
                }]
            })
        }
        internalCache.RoomCharacterList.set({
            key: RoomId,
            value: remainingCharacters
        })

    }, { retryErrors: ['TransactionCanceledException']})
}

export const disconnectMessage = async ({ payloads }: { payloads: DisconnectMessage[], messageBus?: MessageBus }): Promise<void> => {
    //
    // TODO: Figure out whether a forced disconnet invalidates any cached values
    //

    await Promise.all(payloads.map(async (payload) => {
        const ConnectionId = `CONNECTION#${payload.connectionId}`
        const characterQuery = await connectionDB.query({
            ConnectionId,
            ExpressionAttributeValues: {
                ':dcPrefix': 'CHARACTER#'
            },
            KeyConditionExpression: 'begins_with(DataCategory, :dcPrefix)',
            ProjectionFields: ['DataCategory']
        })
        await Promise.all([
            ...characterQuery.map(async ({ DataCategory }) => (atomicallyRemoveCharacterAdjacency(payload.connectionId, splitType(DataCategory)[1]))),
            connectionDB.deleteItem({
                ConnectionId,
                DataCategory: 'Meta::Connection'
            }),
            connectionDB.optimisticUpdate({
                key: {
                    ConnectionId: 'Global',
                    DataCategory: 'Connections'
                },
                updateKeys: ['connections'],
                updateReducer: (draft) => {
                    draft.connections[payload.connectionId] = undefined
                }
            }),
            connectionDB.optimisticUpdate({
                key: {
                    ConnectionId: 'Library',
                    DataCategory: 'Subscriptions'
                },
                updateKeys: ['ConnectionIds'],
                updateReducer: (draft) => {
                    draft.ConnectionIds = draft.ConnectionIds.filter((value) => (value !== payload.connectionId))
                }
            })
        ])
    }))
}

export default disconnectMessage
