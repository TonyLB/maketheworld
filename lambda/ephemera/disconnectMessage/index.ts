import { DisconnectMessage, MessageBus } from "../messageBus/baseClasses"

import { connectionDB, ephemeraDB, exponentialBackoffWrapper, multiTableTransactWrite } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { marshall } from "@aws-sdk/util-dynamodb"
import { splitType } from "@tonylb/mtw-utilities/dist/types"
import messageBus from "../messageBus"
import internalCache from "../internalCache"
import { EphemeraCharacterId } from "@tonylb/mtw-interfaces/dist/baseClasses"

const atomicallyRemoveCharacterAdjacency = async (connectionId: string, characterId: EphemeraCharacterId) => {
    return exponentialBackoffWrapper(async () => {
        const [currentConnections, characterFetch, mapSubscriptions] = await Promise.all([
            internalCache.CharacterConnections.get(characterId),
            internalCache.CharacterMeta.get(characterId),
            internalCache.Global.get("mapSubscriptions")
        ])
        if (!(currentConnections && currentConnections.length)) {
            return
        }
        const { RoomId, Name } = characterFetch || {}
        const { Pronouns, assets, ...rest } = characterFetch || {}

        const currentActiveCharacters = await internalCache.RoomCharacterList.get(RoomId)

        const remainingConnections = currentConnections.filter((value) => (value !== connectionId))

        const remainingCharacters = [
            ...(currentActiveCharacters || []).filter(({ EphemeraId }) => (EphemeraId !== characterId)),
            ...((remainingConnections.length > 0)
                ? [{
                    ...rest,
                    ConnectionIds: remainingConnections
                }]
                : []
            )
        ]
        const adjustMeta = remainingConnections.length > 0
            ? [{
                Update: {
                    TableName: 'Connections',
                    Key: marshall({
                        ConnectionId: characterId,
                        DataCategory: 'Meta::Character'
                    }),
                    UpdateExpression: 'SET connections = :newConnections',
                    ExpressionAttributeValues: marshall({
                        ':newConnections': remainingConnections,
                        ':oldConnections': currentConnections
                    }),
                    ConditionExpression: 'connections = :oldConnections'
                }
            }]
            : [{
                Delete: {
                    TableName: 'Connections',
                    Key: marshall({
                        ConnectionId: characterId,
                        DataCategory: 'Meta::Character'
                    }),
                }
            }]
        const adjustMapSubscriptions = (mapSubscriptions || []).find(({ connectionId: check }) => (check === connectionId))
            ? [{
                Update: {
                    TableName: 'Connections',
                    Key: marshall({
                        ConnectionId: 'Map',
                        DataCategory: 'Subscriptions'
                    }),
                    UpdateExpression: 'SET connections = :newConnections',
                    ExpressionAttributeValues: marshall({
                        ':newConnections': (mapSubscriptions || []).filter(({ connectionId: check }) => (check !== connectionId)),
                        ':oldConnections': mapSubscriptions
                    }),
                    ConditionExpression: 'connections = :oldConnections'
                }
            }]
            : []
        await multiTableTransactWrite([{
            Delete: {
                TableName: 'Connections',
                Key: marshall({
                    ConnectionId: `CONNECTION#${connectionId}`,
                    DataCategory: characterId
                })
            }
        },
        ...adjustMapSubscriptions,
        ...adjustMeta,
        {
            Update: {
                TableName: 'Ephemera',
                Key: marshall({
                    EphemeraId: RoomId,
                    DataCategory: 'Meta::Room'
                }),
                UpdateExpression: 'SET activeCharacters = :newCharacters',
                ExpressionAttributeValues: marshall({
                    ':newCharacters': remainingCharacters,
                    ':oldCharacters': currentActiveCharacters
                }),
                ConditionExpression: 'activeCharacters = :oldCharacters'
            }
        }])
        if (remainingConnections.length === 0) {
            messageBus.send({
                type: 'EphemeraUpdate',
                updates: [{
                    type: 'CharacterInPlay',
                    CharacterId: characterId,
                    Connected: false,
                    targets: ['GLOBAL', `!CONNECTION#${connectionId}`]
                }]
            })
            messageBus.send({
                type: 'PublishMessage',
                targets: [RoomId, `!${characterId}`],
                displayProtocol: 'WorldMessage',
                message: [{
                    tag: 'String',
                    value: `${Name || 'Someone'} has disconnected.`
                }]
            })
            messageBus.send({
                type: 'RoomUpdate',
                roomId: RoomId
            })
        }
        internalCache.RoomCharacterList.set({
            key: RoomId,
            value: remainingCharacters
        })

    }, { retryErrors: ['TransactionCanceledException']})
}

export const disconnectMessage = async ({ payloads }: { payloads: DisconnectMessage[], messageBus?: MessageBus }): Promise<void> => {
    //
    // TODO: Figure out whether a forced disconnet invalidates any cached values
    //

    await Promise.all(payloads.map(async (payload) => {
        const ConnectionId = `CONNECTION#${payload.connectionId}`
        const characterQuery = await connectionDB.query<{ DataCategory: EphemeraCharacterId }[]>({
            ConnectionId,
            ExpressionAttributeValues: {
                ':dcPrefix': 'CHARACTER#'
            },
            KeyConditionExpression: 'begins_with(DataCategory, :dcPrefix)',
            ProjectionFields: ['DataCategory']
        })
        await Promise.all([
            ...characterQuery.map(({ DataCategory }) => (atomicallyRemoveCharacterAdjacency(payload.connectionId, DataCategory))),
            connectionDB.deleteItem({
                ConnectionId,
                DataCategory: 'Meta::Connection'
            }),
            connectionDB.optimisticUpdate({
                key: {
                    ConnectionId: 'Global',
                    DataCategory: 'Connections'
                },
                updateKeys: ['connections'],
                updateReducer: (draft) => {
                    draft.connections[payload.connectionId] = undefined
                }
            }),
            connectionDB.optimisticUpdate({
                key: {
                    ConnectionId: 'Library',
                    DataCategory: 'Subscriptions'
                },
                updateKeys: ['ConnectionIds'],
                updateReducer: (draft) => {
                    draft.ConnectionIds = draft.ConnectionIds.filter((value) => (value !== payload.connectionId))
                }
            })
        ])
    }))
}

export default disconnectMessage
