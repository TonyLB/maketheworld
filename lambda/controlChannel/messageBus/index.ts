import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'

import {
    MessageType,
    isPublishMessage,
    isEphemeraUpdate,
    isDisconnectMessage,
    isConnectMessage,
    isWhoAmIMessage,
    isSyncRequest,
    isSyncResponse,
    isRegisterCharacterMessage
} from "./baseClasses"

import publishMessage from '../publishMessage'
import ephemeraUpdate from '../ephemeraUpdate'
import disconnectMessage from '../disconnectMessage'
import connectMessage from '../connectMessage'
import whoAmIMessage from '../whoAmI'
import { syncRequest, syncResponse } from '../syncHandler'
import registerCharacter from '../registerCharacter'

export class MessageBus extends InternalMessageBus<MessageType> {
    constructor() {
        super()
        this.subscribe({
            tag: 'PublishMessage',
            priority: 5,
            filter: isPublishMessage,
            callback: publishMessage
        })
        this.subscribe({
            tag: 'EphemeraUpdate',
            priority: 5,
            filter: isEphemeraUpdate,
            callback: ephemeraUpdate
        })
        this.subscribe({
            tag: 'Disconnect',
            priority: 1,
            filter: isDisconnectMessage,
            callback: disconnectMessage
        })
        this.subscribe({
            tag: 'Connect',
            priority: 1,
            filter: isConnectMessage,
            callback: connectMessage
        })
        this.subscribe({
            tag: 'WhoAmI',
            priority: 1,
            filter: isWhoAmIMessage,
            callback: whoAmIMessage
        })
        this.subscribe({
            tag: 'SyncResponse',
            priority: 2,
            filter: isSyncResponse,
            callback: syncResponse
        })
        this.subscribe({
            tag: 'SyncRequest',
            priority: 3,
            filter: isSyncRequest,
            callback: syncRequest
        })
        this.subscribe({
            tag: 'RegisterCharacter',
            priority: 1,
            filter: isRegisterCharacterMessage,
            callback: registerCharacter
        })
    }
}

export const messageBus = new MessageBus()
export default messageBus