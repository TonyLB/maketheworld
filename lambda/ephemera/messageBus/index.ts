import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'

import {
    MessageBus,
    isPublishMessage,
    isEphemeraUpdate,
    isDisconnectMessage,
    isConnectMessage,
    isWhoAmIMessage,
    isSyncRequest,
    isSyncResponse,
    isRegisterCharacterMessage,
    isFetchPlayerEphemera,
    isImportDefaults,
    isFetchImportDefaults,
    isPerception,
    isMoveCharacter,
    isDecacheAsset,
    isCacheAssetMessage,
    isPlayerUpdateMessage,
    isRoomUpdateMessage,
    isDescentUpdateMessage,
    isAncestryUpdateMessage,
    isDependencyCascadeMessage,
    isExecuteActionMessage,
    isMapSubscription,
    isMapUpdateMessage
} from "./baseClasses"

import publishMessage from '../publishMessage'
import ephemeraUpdate from '../ephemeraUpdate'
import disconnectMessage from '../disconnectMessage'
import connectMessage from '../connectMessage'
import whoAmIMessage from '../whoAmI'
import { syncRequest, syncResponse } from '../syncHandler'
import registerCharacter from '../registerCharacter'
import { fetchPlayerEphemera } from '../fetchEphemera'
import perceptionMessage from '../perception'
import moveCharacter from '../moveCharacter'
import decacheAssetMessage from '../decacheMessage'
import { cacheAssetMessage } from '../cacheAsset'
import playerUpdateMessage from '../playerUpdate'
import roomUpdateMessage from '../roomUpdate'
import dependentUpdateMessage from '../dependentMessages/dependentUpdate'
import dependencyCascadeMessage from '../dependentMessages/dependencyCascade'
import executeActionMessage from '../executeAction'
import mapSubscriptionMessage from '../mapSubscription'
import mapUpdateMessage from '../mapUpdate'

export const messageBus = new MessageBus()
messageBus.subscribe({
    tag: 'PublishMessage',
    priority: 15,
    filter: isPublishMessage,
    callback: publishMessage
})
messageBus.subscribe({
    tag: 'EphemeraUpdate',
    priority: 10,
    filter: isEphemeraUpdate,
    callback: ephemeraUpdate
})
messageBus.subscribe({
    tag: 'Disconnect',
    priority: 1,
    filter: isDisconnectMessage,
    callback: disconnectMessage
})
messageBus.subscribe({
    tag: 'Connect',
    priority: 1,
    filter: isConnectMessage,
    callback: connectMessage
})
messageBus.subscribe({
    tag: 'WhoAmI',
    priority: 1,
    filter: isWhoAmIMessage,
    callback: whoAmIMessage
})
messageBus.subscribe({
    tag: 'SyncResponse',
    priority: 2,
    filter: isSyncResponse,
    callback: syncResponse
})
messageBus.subscribe({
    tag: 'SyncRequest',
    priority: 3,
    filter: isSyncRequest,
    callback: syncRequest
})
messageBus.subscribe({
    tag: 'RegisterCharacter',
    priority: 1,
    filter: isRegisterCharacterMessage,
    callback: registerCharacter
})
messageBus.subscribe({
    tag: 'FetchPlayerEphemera',
    priority: 2,
    filter: isFetchPlayerEphemera,
    callback: fetchPlayerEphemera
})
messageBus.subscribe({
    tag: 'MapSubscription',
    priority: 2,
    filter: isMapSubscription,
    callback: mapSubscriptionMessage
})
messageBus.subscribe({
    tag: 'Perception',
    priority: 10,
    filter: isPerception,
    callback: perceptionMessage
})
messageBus.subscribe({
    tag: 'MoveCharacter',
    priority: 4,
    filter: isMoveCharacter,
    callback: moveCharacter
})
messageBus.subscribe({
    tag: 'DecacheAsset',
    priority: 1,
    filter: isDecacheAsset,
    callback: decacheAssetMessage
})
messageBus.subscribe({
    tag: 'CacheAsset',
    priority: 1,
    filter: isCacheAssetMessage,
    callback: cacheAssetMessage
})
messageBus.subscribe({
    tag: 'UpdatePlayer',
    priority: 5,
    filter: isPlayerUpdateMessage,
    callback: playerUpdateMessage
})
messageBus.subscribe({
    tag: 'RoomUpdate',
    priority: 6,
    filter: isRoomUpdateMessage,
    callback: roomUpdateMessage
})
messageBus.subscribe({
    tag: 'MapUpdate',
    priority: 15,
    filter: isMapUpdateMessage,
    callback: mapUpdateMessage
})
messageBus.subscribe({
    tag: 'AncestryUpdate',
    priority: 3,
    filter: isAncestryUpdateMessage,
    callback: dependentUpdateMessage('Ancestry')
})
messageBus.subscribe({
    tag: 'DescentUpdate',
    priority: 4,
    filter: isDescentUpdateMessage,
    callback: dependentUpdateMessage('Descent')
})

messageBus.subscribe({
    tag: 'DependencyCascade',
    priority: 5,
    filter: isDependencyCascadeMessage,
    callback: dependencyCascadeMessage
})

messageBus.subscribe({
    tag: 'ExecuteAction',
    priority: 7,
    filter: isExecuteActionMessage,
    callback: executeActionMessage
})

export default messageBus