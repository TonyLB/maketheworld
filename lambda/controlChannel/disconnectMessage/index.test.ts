jest.mock('@tonylb/mtw-utilities/dist/dynamoDB/index')
import {
    ephemeraDB,
    connectionDB,
    multiTableTransactWrite
} from '@tonylb/mtw-utilities/dist/dynamoDB/index'

jest.mock('../messageBus')
import messageBus from '../messageBus'

const ephemeraDBMock = ephemeraDB as jest.Mocked<typeof ephemeraDB>
const connectionDBMock = connectionDB as jest.Mocked<typeof connectionDB>
const multiTableTransactWriteMock = multiTableTransactWrite as jest.Mock
const messageBusMock = messageBus as jest.Mocked<typeof messageBus>

import disconnectMessage from '.'

describe("disconnectMessage", () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
    })

    it("should update correctly on last connection", async () => {
        ephemeraDBMock.getItem.mockResolvedValueOnce({
            RoomId: 'TestABC'
        })
        .mockResolvedValueOnce({
            activeCharacters: [
                {
                    EphemeraId: 'CHARACTER#BCD',
                    Name: 'TestToo',
                    Connections: ['BCD']
                },
                {
                    EphemeraId: 'CHARACTER#ABC',
                    Name: 'Tess',
                    Connections: ['XYZ']
                }
            ]
        })
        connectionDBMock.getItem.mockResolvedValueOnce({
            connections: ['XYZ']
        })
        connectionDBMock.query.mockResolvedValueOnce([{ DataCategory: 'CHARACTER#ABC' }])
        await disconnectMessage({
            payloads: [{ type: 'Disconnect', connectionId: 'XYZ' }],
            messageBus
        })
        expect(multiTableTransactWrite).toHaveBeenCalledTimes(1)
        expect(multiTableTransactWriteMock.mock.calls[0][0]).toMatchSnapshot()
    })

    it("should update correctly on redundant connections", async () => {
        ephemeraDBMock.getItem.mockResolvedValueOnce({
            RoomId: 'TestABC'
        })
        .mockResolvedValueOnce({
            activeCharacters: [
                {
                    EphemeraId: 'CHARACTER#BCD',
                    Name: 'TestToo',
                    Connections: ['BCD']
                },
                {
                    EphemeraId: 'CHARACTER#ABC',
                    Name: 'Tess',
                    Connections: ['QRS', 'XYZ']
                }
            ]
        })
        connectionDBMock.getItem.mockResolvedValueOnce({
            connections: ['QRS', 'XYZ']
        })
        connectionDBMock.query.mockResolvedValueOnce([{ DataCategory: 'CHARACTER#ABC' }])
        await disconnectMessage({
            payloads: [{ type: 'Disconnect', connectionId: 'XYZ' }],
            messageBus
        })
        expect(multiTableTransactWrite).toHaveBeenCalledTimes(1)
        expect(multiTableTransactWriteMock.mock.calls[0][0]).toMatchSnapshot()
    })

})