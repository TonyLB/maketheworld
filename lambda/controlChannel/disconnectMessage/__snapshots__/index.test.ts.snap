// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disconnectMessage should update correctly on last connection 1`] = `
Array [
  Object {
    "Delete": Object {
      "Key": Object {
        "ConnectionId": Object {
          "S": "CONNECTION#XYZ",
        },
        "DataCategory": Object {
          "S": "CHARACTER#ABC",
        },
      },
      "TableName": "Connections",
    },
  },
  Object {
    "Delete": Object {
      "Key": Object {
        "ConnectionId": Object {
          "S": "CHARACTER#ABC",
        },
        "DataCategory": Object {
          "S": "Meta::Character",
        },
      },
      "TableName": "Connections",
    },
  },
  Object {
    "Update": Object {
      "ConditionExpression": "activeCharacters = :oldCharacters",
      "ExpressionAttributeValues": Object {
        ":newCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "BCD",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
          ],
        },
        ":oldCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "BCD",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "XYZ",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "Name": Object {
                  "S": "Tess",
                },
              },
            },
          ],
        },
      },
      "Key": Object {
        "DataCategory": Object {
          "S": "Meta::Room",
        },
        "EphemeraId": Object {
          "S": "ROOM#TestABC",
        },
      },
      "TableName": "Ephemera",
      "UpdateExpression": "SET activeCharacters = :newCharacters",
    },
  },
]
`;

exports[`disconnectMessage should update correctly on redundant connections 1`] = `
Array [
  Object {
    "Delete": Object {
      "Key": Object {
        "ConnectionId": Object {
          "S": "CONNECTION#XYZ",
        },
        "DataCategory": Object {
          "S": "CHARACTER#ABC",
        },
      },
      "TableName": "Connections",
    },
  },
  Object {
    "Update": Object {
      "ConditionExpression": "connections = :oldConnections",
      "ExpressionAttributeValues": Object {
        ":newConnections": Object {
          "L": Array [
            Object {
              "S": "QRS",
            },
          ],
        },
        ":oldConnections": Object {
          "L": Array [
            Object {
              "S": "QRS",
            },
            Object {
              "S": "XYZ",
            },
          ],
        },
      },
      "Key": Object {
        "ConnectionId": Object {
          "S": "CHARACTER#ABC",
        },
        "DataCategory": Object {
          "S": "Meta::Character",
        },
      },
      "TableName": "Connections",
      "UpdateExpression": "SET connections = :newConnections",
    },
  },
  Object {
    "Update": Object {
      "ConditionExpression": "activeCharacters = :oldCharacters",
      "ExpressionAttributeValues": Object {
        ":newCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "BCD",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "QRS",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "HomeId": Object {
                  "S": "VORTEX",
                },
                "Name": Object {
                  "S": "Tess",
                },
                "RoomId": Object {
                  "S": "TestABC",
                },
              },
            },
          ],
        },
        ":oldCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "BCD",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "QRS",
                    },
                    Object {
                      "S": "XYZ",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "Name": Object {
                  "S": "Tess",
                },
              },
            },
          ],
        },
      },
      "Key": Object {
        "DataCategory": Object {
          "S": "Meta::Room",
        },
        "EphemeraId": Object {
          "S": "ROOM#TestABC",
        },
      },
      "TableName": "Ephemera",
      "UpdateExpression": "SET activeCharacters = :newCharacters",
    },
  },
]
`;
